src/streamlink_cli/main.py

```diff
@@ -375,6 +397,13 @@ def read_stream(stream, output, prebuffer, chunk_size=8192):
                     console.exit("Error when writing to output: {0}, exiting", err)
 
                 break
+            except ValueError as err:
+                try:
+                    output.player.wait(10) # Wait a moment before we check the player
+                    if not output.opened:
+                        sys.exit(130)
+                except TimeoutError:
+                    console.exit("Error when writing to output: {0}, exiting", err)
```

Comment:
For me it looks like the handler and the main thread are having some kind of race condition causing the issue with the exception occurring during the write I mentioned in the OP. This causes the player to be closed but the check for ```player.opened``` still returning ```True```. 

My thought process was that the easiest solution would be to catch the Exception and wait a few seconds to see if the player is closing. If yes it is assumed that the exception was caused by the race condition during the exit and not an actual error if not the usual error message is displayed. The first case should only occur if streamlink was interrupted with CTRL-C that's why I'm using the 130 exit code.

There is probably a better solution so let me know :)